### ISSUE TITLE
Edit command is slow

### ISSUE THREAD

**Issue by iwiznia** [Expensify/App#13742]
Context https://expensify.slack.com/archives/C03TQ48KC/p1671566294965399

# Problem

`UpdateComment` and really `Report_EditComment` too (they use the same code) are slow.
Mostly, it is caused by fully loading and instantiating all the report history. 

# Solution

Instead of loading the whole history, to then call `ReportUtils::getMostRecentVisibleAction` and `ReportUtils::getMostRecentVisibleUnreadAction`, we should have auth return those values by querying them directly in the DB and only returning the data we need.

cc @chiragsalian

**Comment by luacmartins** [comment]
So it seems like we want to create a new Auth command, e.g. `getMostRecentVisibleAction` that will return the most recent visible action or the most recent visible unread action to PHP so we can stop calling `$report->getHistory()`. Does that sound right @iwiznia?

**Comment by iwiznia** [comment]
Yeah maybe... I am not that familiar with this flow, like why do we care about the most recent visible action?

**Comment by luacmartins** [comment]
From what I can tell, it seems to be used to:

1. Update the last message text in the LHN if we edited the most recent message
![Screenshot 2022-12-22 at 2 02 36 PM](https://user-images.githubusercontent.com/22219519/209225620-1391864d-818c-4605-b602-90d426e1d9af.png)

3. Update the unread count if we deleted the last unread message

**Comment by luacmartins** [comment]
Maybe we don't need a new command and we can move that logic to `UpdateReportComment` in Auth. I'll explore that first since it would most align with our 1:1:1 goal.

**Comment by iwiznia** [comment]
The last 1 of `1:1:1` was not decided and I expressed my concern about it many times before: adding reads to a write makes the command more likely to conflict, so I am not sure if it's best to add it in that command or have a separate one.

**Comment by luacmartins** [comment]
Makes sense, I think in that case a new command is better.

**Comment by luacmartins** [comment]
I'm going ooo until Jan 3, so I'll unassign myself in case someone else wants to work on this while I'm away. 

I have some draft PRs ([Web-E](https://github.com/Expensify/Web-Expensify/pull/35902) and [Auth](https://github.com/Expensify/Auth/pull/7363)), but they are far from being ready. If anyone picks this up, feel free to use/discard my draft PRs. If nobody picks this up, I'll come back to it in January!

**Comment by JmillsExpensify** [comment]
Thanks for the heads up! I'm trying to wrap my head around the importance of this. My initial sense is that it's importance for Expensify employees but not particularly important for WAQ. So we should solve it but it's not critically pressing. Does that sound right to others on this issue?

**Comment by JmillsExpensify** [comment]
Re-added the bug label so that we can get someone else on this over the holidays. And....that didn't work. Going with the engineering label instead.

**Comment by JmillsExpensify** [comment]
@sketchydroide Do you think you'll have time to help move this forward this week? Carlos has [two draft PRs](https://github.com/Expensify/App/issues/13742#issuecomment-1364249475) up as a starting point.

**Comment by sketchydroide** [comment]
Yeah, I can help with this, if I get stuck, I'll be sure to ask for help, but for now I think I understand the core of it

**Comment by JmillsExpensify** [comment]
Awesome, thanks!

**Comment by sketchydroide** [comment]
I've started work on it, kept the changes of Carlos, and working on top of the original PRs
I think I have a good grasp of what is needed, I think most of the changes wil be on Auth and then Web is mostly cleanup, although I think I need to create a command for `getMostRecentVisibleUnreadAction` both on Auth and Web.

This is the one that used the lastReadSequenceNumber that I will not use, will use the timeStamp instead as per the new deprecate sequenceNumber doc.

**Comment by sketchydroide** [comment]
>Thanks for the heads up! I'm trying to wrap my head around the importance of this. My initial sense is that it's importance for Expensify employees but not particularly important for WAQ. So we should solve it but it's not critically pressing. Does that sound right to others on this issue?

I think it's important for anyone that uses the app, unless only us can edit the comments, and if that is true why? It works pretty well, and it's a great feature.
Am I missing something?
Not sure if WAQ though, as it's not exactly a bug but improving a feature... It can be a thin line though 🤷🏼

**Comment by iwiznia** [comment]
This is definitely WAQ

**Comment by iwiznia** [comment]
I agree. This is important for anyone using the app, editing a comment is such a basic action.

**Comment by JmillsExpensify** [comment]
Oh ooops. I didn't mean for that comment to hit on this issue, my b. This was meant for a chronos bug.

**Comment by sketchydroide** [comment]
I think the Auth PR is close, I need to add integration tests
I played a bit with the queries in my local DB and they gave me the results I wanted, so they seem ok.
After the integration tests and wrapping up the Wep PR accordingly I will do so some local testing see if works properly

**Comment by JmillsExpensify** [comment]
Awesome, thanks for moving this forward over the holidays!

**Comment by sketchydroide** [comment]
Continue to work on this, moslty the tests for today, I hope to have the tests done today

**Comment by sketchydroide** [comment]
Still working on the Tests, althoug for one of the commands `GetMostRecentVisibleUnreadAction` I need some help from the people working on sequenceNumbers deprecation as it ovelaps.

If I get no input we can break the 2 commands apart in 2 different PRs (2 Auth and 2 Web) for each command

**Comment by JmillsExpensify** [comment]
Following up that, do you think you'll be ready to open up your PRs for reviews this week?

**Comment by sketchydroide** [comment]
I think so, I'm getting much closer, I think I need to discuss a bit how this is affected by sequenceNumber refactoring, since there is some work arounf unread messages

**Comment by sketchydroide** [comment]
I think I've done as much as I can for now, [I had another chat with @roryabraham](https://expensify.slack.com/archives/C03TQ48KC/p1672763079958549?thread_ts=1672753087.553539&cid=C03TQ48KC), and we concluded that because some of the logic is dependent, would need changes after, or would be deprecated with the refactor of the sequence numbers, that we should put this on **hold** until a couple of PRs are done.
Namely:
https://github.com/Expensify/App/pull/13419
https://github.com/Expensify/Expensify/issues/240913

**Comment by sketchydroide** [comment]
still on hold, same reasons
https://github.com/Expensify/App/pull/13419 has been merged, but the https://github.com/Expensify/Expensify/issues/240913 is still being worked on by Rory

**Comment by sketchydroide** [comment]
still same as above

**Comment by JmillsExpensify** [comment]
Thanks for keeping on this one. I clarified the remaining PR we're holding on in the title, but otherwise we'll hopefully be unblocked in the coming weeks.

**Comment by JmillsExpensify** [comment]
Still held on the linked issue above.

**Comment by JmillsExpensify** [comment]
Same same.

**Comment by sketchydroide** [comment]
thanks, I was OOO but can now do the updates again

**Comment by sketchydroide** [comment]
still same

**Comment by sketchydroide** [comment]
From what I understand this is still on the same, it does seem that the last GH holding this is getting closer to completion.
@roryabraham does that sound true?

**Comment by JmillsExpensify** [comment]
Hmm, actually shouldn't we take the hold off now?

**Comment by iwiznia** [comment]
I implemented a quick hack here https://github.com/Expensify/Web-Expensify/pull/36315 for AddComment. Maybe we can implement something similar for this, then work on a better solution that works for both.

**Comment by roryabraham** [comment]
Yep we can take this off hold

**Comment by sketchydroide** [comment]
Great, removing hold.
@roryabraham I might need to chat with you again to check on the logic to use again.
Will see if I can make sense of it first.

**Comment by sketchydroide** [comment]
Reviewing my previous changes, and looking to see what logic has been chnaged since this was on hold.
Then will determin if the changes I did before still make sense.
I don't think I will finish it this week, but early next week seems doable.

**Comment by iwiznia** [comment]
My bet is that the worst offender is that we are calling `$chatReport->getHistory()` without params, so we need to avoid doing that, since for big chats that takes forever.
LMK if you need some help with this.

**Comment by sketchydroide** [comment]
I think I've got a good grasp of what I need to do.
It would be great for me to know your thoughts on the queries I created on `Report.cpp` though. PR [here](https://github.com/Expensify/Auth/pull/7363)

**Comment by sketchydroide** [comment]
Thanks @iwiznia for the comments on the PR, will go through them today

**Comment by sketchydroide** [comment]
Auth PR is done, and I will put it for review tomorrow
Web should be very close, it's much smaller, it might need some work on the App

**Comment by sketchydroide** [comment]
ok [Auth PR](https://github.com/Expensify/Auth/pull/7363) is now ready for review

**Comment by JmillsExpensify** [comment]
Looks like we're still working through the linked Auth PR.

**Comment by sketchydroide** [comment]
It's under review, and there has been some comments on what can be improved, and some doubts poping up
I'll get them sorted today, and maybe get a convo happening if it becomes to complex for a PR.

**Comment by sketchydroide** [comment]
Auth PR seems to be close to being done with the comments

**Comment by JmillsExpensify** [comment]
Assigning Ionatan on this as he takes over the linked PR.

**Comment by MelvinBot** [comment]
@iwiznia, @JmillsExpensify Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by iwiznia** [comment]
Waiting for reviews

**Comment by MelvinBot** [comment]
@iwiznia, @JmillsExpensify Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by MelvinBot** [comment]
@iwiznia, @JmillsExpensify Still overdue 6 days?! Let's take care of this!

**Comment by iwiznia** [comment]
Working on it, conflicts and people reusing a method I had killed delayed this a while, plus it will require waiting for an auth deploy.

**Comment by MelvinBot** [comment]
@iwiznia, @JmillsExpensify Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by MelvinBot** [comment]
@iwiznia, @JmillsExpensify 6 days overdue. This is scarier than being forced to listen to Vogon poetry!

**Comment by MelvinBot** [comment]
@iwiznia, @JmillsExpensify Now this issue is 8 days overdue. Are you sure this should be a Daily? Feel free to change it!

**Comment by MelvinBot** [comment]
@iwiznia, @JmillsExpensify 10 days overdue. I'm getting more depressed than Marvin.

**Comment by JmillsExpensify** [comment]
Still in progress. Thanks Melvin

**Comment by MelvinBot** [comment]
@iwiznia, @JmillsExpensify Huh... This is 4 days overdue. Who can take care of this?

**Comment by MelvinBot** [comment]
@iwiznia, @JmillsExpensify Still overdue 6 days?! Let's take care of this!
